### PR TITLE
Refactor connection handler

### PR DESCRIPTION
**Pull request by arthurnn**
~~~**NOTICE** This is still in early stages, but I want to create a PR, so we can discuss based on some concrete implementation that have been discussed for a while among the Rails team.~~~
### Summary

ConnectionHandler will not have any knowledge of AR models now, it will
only know about the specs.
Like that we can decouple the two, and allow the same model to use more
than one connection.
### Context

Historically, folks used to create abstract AR classes on the fly in
order to have multiple connections for the same model, and override the
connection methods.

With this, now we can override the `specificiation_id` method in the
model, to return a key, that will be used to find the connection_pool
from the handler.

@jeremy thoughts?
### How can users extend this?

database.yml

```
production:
  database: ...
readonly:
  database: ...
```

```
class User < ApplicationRecord
  self.connection_specification_name = "readonly"
end
```

or

```
class User < ApplicationRecord
  def self.connection_specification_name 
     @in_readonly_mode ? "readonly" : "primary"
  end
end
```

In this example `in_readonly_mode` could be an ivar that is set at runtime to change the connection at runtime.

Pretty much the gist of this is, models can now extend that method, and change the connection it will be used. This is possible because now the connection handler doesnt hash the connections by the class name anymore.
### Is this backwards compatible?

Yes, 
- The default `connection_specification_name` behaviour is to fallback to the parent method, if `connection_specification_name` not defined.
- When doing an establish_connection in a AR model: ie: `User.establish_connection()`, that will set the `connection_specification_name` to the model name.

### CODE REVIEW

File: activerecord/lib/active_record/connection_handling.rb [hunk 1]
```diff
@@ -87,6 +90,23 @@ def connection
...
+      find_legacy_spec_id(self) if fallback
+    end
+
+    def find_legacy_spec_id(klass)
+      return "primary" if klass == Base
+      if id = klass.specification_id(false)
+        return id
+      end
+      find_legacy_spec_id(klass.superclass)
+    end
```
Comment (arthurnn) [inline activerecord/lib/active_record/connection_handling.rb]: We need to do some gymnastics here to be backwards compatible.
However if we deprecate the `establish_connection` from the model, eventually we would be able to remove this fallback into the `Class.name` method.

The way I envision this, is we would establish the connection straight to the handler, as this connection is not attached to the model anymore.

Comment (arthurnn) [inline activerecord/lib/active_record/connection_handling.rb]: Ignore this comment.

File: activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb [hunk 2]
```diff
@@ -927,28 +914,24 @@ def owner_to_pool
...
+            owner_to_pool[spec_id] = nil
           end
         }
       end
 
-      def pool_from_any_process_for(owner)
-        owner_to_pool = @owner_to_pool.values.find { |v| v[owner.name] }
-        owner_to_pool && owner_to_pool[owner.name]
+      def pool_from_any_process_for(spec_id)
+        owner_to_pool = @owner_to_pool.values.find { |v| v[spec_id] }
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb]: Could spare an array allocation by calling `each_value` if `Concurrent::Map` supports it 😁

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb [hunk 3]
```diff
@@ -3,10 +3,10 @@
 module ActiveRecord
   module ConnectionAdapters
     class ConnectionSpecification #:nodoc:
-      attr_reader :config, :adapter_method
+      attr_reader :config, :adapter_method, :id
 
-      def initialize(config, adapter_method)
-        @config, @adapter_method = config, adapter_method
+      def initialize(id, config, adapter_method)
+        @config, @adapter_method, @id = config, adapter_method, id
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: Readability: find the `id` shifting from being the last arg to first here confusing. I would keep it consistent by always putting it first.

File: activerecord/lib/active_record/connection_handling.rb [hunk 4]
```diff
@@ -87,6 +91,19 @@ def connection
       retrieve_connection
     end
 
+    def specification_id=(value)
+      @specification_id = value
+    end
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_handling.rb]: `attr_writer :specification_id`?

File: activerecord/lib/active_record/connection_handling.rb [hunk 5]
```diff
@@ -87,6 +91,19 @@ def connection
...
+
+    # Return the specification id from this class otherwise look it up
+    # in the parent.
+    def specification_id
+      unless defined?(@specification_id)
+        @specification_id = self == Base ? "primary" : superclass.specification_id
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_handling.rb]: Won't `super` do instead of `superclass.specification_id`?

File: activerecord/lib/active_record/connection_handling.rb [hunk 6]
```diff
@@ -45,16 +45,20 @@ module ConnectionHandling
     # The exceptions AdapterNotSpecified, AdapterNotFound and +ArgumentError+
     # may be returned on an error.
     def establish_connection(spec = nil)
+      raise RuntimeError, "Anonymous class is not allowed." unless name
+
       spec     ||= DEFAULT_ENV.call.to_sym
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_handling.rb]: Having two variables clash name but differ on semantics is confusing. Maybe renaming this to `spec_name` clears it up a bit?

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb [hunk 7]
```diff
@@ -164,7 +164,7 @@ def resolve_all
         #   spec.config
         #   # => { "host" => "localhost", "database" => "foo", "adapter" => "sqlite3" }
         #
-        def spec(config)
+        def spec(config, id = nil)
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: Find the use of `id` throughout confusing. Why not go with `specification_name`? That feels much closer to its nature of being symbols and strings like `:production`, `"primary"`. Put another way id reads too abstract to me, while name is more down to earth :smile:

File: activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb [hunk 8]
```diff
@@ -910,15 +904,8 @@ def remove_connection(owner)
...
-        class_to_pool[klass.name] ||= begin
-          until pool = pool_for(klass)
-            klass = klass.superclass
-            break unless klass <= Base
-          end
-
-          class_to_pool[klass.name] = pool
-        end
+      def retrieve_connection_pool(spec_id)
+        pool_for(spec_id)
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb]: Should we just inline `pool_for` or is it used elsewhere?

File: activerecord/lib/active_record/connection_handling.rb [hunk 9]
```diff
@@ -45,16 +45,20 @@ module ConnectionHandling
...
       resolver =   ConnectionAdapters::ConnectionSpecification::Resolver.new configurations
-      spec     =   resolver.spec(spec)
+      # TODO: uses name on establish_connection, for backwards compatibility
+      spec     =   resolver.spec(spec, self == Base ? "primary" : name)
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_handling.rb]: Should we use `===` here for the comparison and not `==`? Think they map to `kind_of?` and `is_a?`, respectively, and that we don't want subclasses to hit the primary spec.

Also is this the reason for the optional `id` param in `spec`? Don't like changing the public API for this conditional.

Comment (kaspth) [inline activerecord/lib/active_record/connection_handling.rb]: Nah, after a quick dip in irb it seems `==` will do fine here. Got confused because I've only ever seen class comparisons done with `===` so thought something was a foot (but everything was in hand 🤓).

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb [hunk 10]
```diff
@@ -179,7 +179,14 @@ def spec(config)
           end
 
           adapter_method = "#{spec[:adapter]}_connection"
-          ConnectionSpecification.new(spec, adapter_method)
+
+          id ||=
+            if config.is_a?(Symbol)
+              config.to_s
+            else
+              "primary"
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: Something really irks me about these two conditionals. Partly that we can't know if the inner symbol check is run at all, or just the special treatment of symbols altogether. Haven't got anything more concrete right now, will let it simmer a bit 😁

File: activerecord/lib/active_record/connection_handling.rb (hunk 5 above)
Comment (arthurnn) [inline activerecord/lib/active_record/connection_handling.rb]: i tried that, but it didnt work. I think because these are class methods. dunno much tbh.

File: activerecord/lib/active_record/connection_handling.rb (hunk 6 above)
Comment (arthurnn) [inline activerecord/lib/active_record/connection_handling.rb]: The problem is that `spec` can be a hash, which is the actual config. So `spec_name` would not work all the time.

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb (hunk 7 above)
Comment (arthurnn) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: I chose `id` because i wanted to be clear this is a unique identifier for the connection. When I think about `name` I doing thing that as been unique.

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb (hunk 10 above)
Comment (arthurnn) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: TBH, I am not super happy about this either. I had to do this, so if you do `handler.establish_connection(:readonly)` it would use that key as the spec_id. but also fallback to "primary" if you pass a config hash instead.

**Comment by arthurnn** [comment]
@tenderlove @matthewd @rafaelfranca would ❤️  your review here too.

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb (hunk 7 above)
Comment (matthewd) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: Agree `id` is awkward. To me `name` is slightly better.. I'm struggling to come up with a particularly compelling alternative, though. :confused:

**Comment by matthewd** [comment]
Doesn't this break the current inheritance behaviour?

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb (hunk 7 above)
Comment (arthurnn) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: Ok , 2 against 1.. I can change to name. =)

**Comment by arthurnn** [comment]
> Doesn't this break the current inheritance behaviour?

It should not. the `specification_id` method fallbacks to the parent if not set. Which brings the same inheritance behaviour as before.

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb [hunk 11]
```diff
@@ -6,7 +6,7 @@ class ConnectionSpecification #:nodoc:
       attr_reader :config, :adapter_method, :id
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: Should we put `id` ahead of the pack here as well?

**Comment by arthurnn** [comment]
@matthewd @kaspth thanks for bearing with me on this. I renamed the key to specification_name.

File: activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb [hunk 12]
```diff
@@ -910,14 +904,18 @@ def remove_connection(owner)
...
       # to optimise for.
-      def retrieve_connection_pool(klass)
-        class_to_pool[klass.name] ||= begin
-          until pool = pool_for(klass)
-            klass = klass.superclass
-            break unless klass <= Base
+      def retrieve_connection_pool(spec_name)
+        owner_to_pool.fetch(spec_name) do
+          if ancestor_pool = pool_from_any_process_for(spec_name)
+            # A connection was established in an ancestor process that must have
```
Comment (vipulnsward) [inline activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb]: may have

File: activerecord/lib/active_record/connection_handling.rb [hunk 13]
```diff
@@ -45,16 +45,20 @@ module ConnectionHandling
     # The exceptions AdapterNotSpecified, AdapterNotFound and +ArgumentError+
     # may be returned on an error.
     def establish_connection(spec = nil)
+      raise RuntimeError, "Anonymous class is not allowed." unless name
```
Comment (vipulnsward) [inline activerecord/lib/active_record/connection_handling.rb]: Is this just to be explicit? raise will still raise `RuntimeError`

Comment (arthurnn) [inline activerecord/lib/active_record/connection_handling.rb]: I dunno, I moved that line from the connection_handler class to here, to keep the behavior, so I didnt wanted to change it.

**Comment by rafaelfranca** [comment]
:+1:

When I said documentation I was talking more about documentation I was talking more about documentation in the code like the one we have in the ConnectionHandler class http://edgeapi.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/ConnectionHandler.html. Right not it is outdated and doesn't talk about this new feature.

**Comment by arthurnn** [comment]
### Future work

My plan is to stop using `name` as `connection_specification_name` when calling `Base.establish_connection`. I think what we should do is deprecate `establish_connection` on AR::Base. I think, instead, we should create a method, maybe something like: `AR::Base.connection_handler.establish_connection_with_config(name, config)`.
Also Rails initializer should use that instead of calling `establish_connection` on AR::Base.

Maybe something like:
database.yml
```
production:
  primary:
    database: ...
  readonly:
    database: ...
  cluster2_write:
    database: ...
```

So, the Rails initializer would go to every top level key under the env, and call `establish_connection_with_config` for each, and the models could just point to one of them.

File: activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb [hunk 14]
```diff
@@ -821,6 +820,10 @@ def checkout_and_verify(c)
     # ConnectionHandler accessible via ActiveRecord::Base.connection_handler.
     # All Active Record models use this handler to determine the connection pool that they
     # should use.
+    #
+    # The ConnectionHandler class is not coupled with the Active models, as it has no knowlodge
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb]: Maybe talk about  `connection_specification_name` in the documentation too?

File: activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb [hunk 15]
```diff
@@ -873,28 +871,27 @@ def clear_all_connections!
       # active or defined connection: if it is the latter, it will be
       # opened and set as the active connection for the class it was defined
       # for (not necessarily the current class).
-      def retrieve_connection(klass) #:nodoc:
-        pool = retrieve_connection_pool(klass)
-        raise ConnectionNotEstablished, "No connection pool for #{klass}" unless pool
+      def retrieve_connection(spec_name) #:nodoc:
+        pool = retrieve_connection_pool(spec_name)
+        raise ConnectionNotEstablished, "No connection pool with id #{spec_name} found." unless pool
```
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb]: with id?

**Comment by mikecmpbll** [comment]
this looks really good! i've been working on something similar to facilitate being able to connect to different databases in different threads, on the same model.

this drops the inflexible class => pool association, which was what was preventing this before. is the following the correct usage of this?

```ruby
class User < ApplicationRecord
  def self.connection_specification_name 
     Thread.current[:spec_name] || "primary"
  end
end

threads = []
10.times do |i|
  threads << Thread.new do
    Thread.current[:spec_name] = "db_shard_#{i}"
    users = User.all # <-- will query db_shard_i
  end
end

threads.each(&:join)
```

**Comment by arthurnn** [comment]
@mikecmpbll thats exactly the idea. Now Model and ConnectionHandler are not coupled, so you can get a pool from the handler using names.
Your code looks right, thats one of the main use cases I had in mind.

Regarding the spec, I am, soon, opening a PR, that will allow the full spec. 

Thanks for taking a look

File: activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb [hunk 16]
```diff
@@ -910,14 +907,18 @@ def remove_connection(owner)
       # #fetch is significantly slower than #[]. So in the nil case, no caching will
       # take place, but that's ok since the nil case is not the common one that we wish
       # to optimise for.
```
Comment (bquorning) [inline activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb]: @arthurnn This comment mentions that `#fetch is significantly slower than #[]`, and previous versions were optimized for the common case, and not caching in the `nil` case.

Shouldn’t this version also use `#[]` instead of `#fetch`?

Comment (bquorning) [inline activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb]: I noticed now that the `owner_to_pool.fetch` I’m seeing comes directly from the inlined `#pool_for` method. I’m unsure whether the `#fetch` here is as slow as the comment says.

Anyhow, either the comment or the code should be changed.

Comment (arthurnn) [inline activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb]: Thanks for the PR

**Comment by rzv01** [comment]
@arthurnn sorry for commenting on a closed issue, but looking through this thread + the changes in the code, I'm not sure I understand how I can make use of these changes.

Basically, my scenario is that I would like to be able to have multiple databases configured at runtime (e.g. new user signs up -> make the Rails app "aware" of a new database -> make a new Pool for that database), and I can switch on each request which database to use.

To my understanding, this PR would make the above use-case possible, including using a multi-threaded server like Puma - it seems though that it eludes me how I can actually define the "specifications" for each DB. Could you please give me some pointers in the right direction? 

To summarize:

1. Having connection pools to multiple DBs - is it thread safe? Can I safely switch between DB specs using `connection_specification_name` per request in a multi-threaded server?

2. *How* do I specify the .. connection specifications? Is it possible to do so while the Rails app is running or does it need to restart? :)

Thanks a lot!

**Comment by arthurnn** [comment]
Hey @rzv01 I am glad you asked. I am planning to write a few blog posts on how people can use this, but I didn't have time yet. I did a talk that touch into this changes too, here is the link: https://youtu.be/-41V0qETxmE?list=PLo3sQWbYZbskTlzH7Ert-FucR7QmIuwE5

> Having connection pools to multiple DBs - is it thread safe? Can I safely switch between DB specs using connection_specification_name per request in a multi-threaded server?

The idea is to not use the same connection pool for different database. Each database should have its own connection pool.

>How do I specify the .. connection specifications? Is it possible to do so while the Rails app is running or does it need to restart? :)

You can create the connection pools passing a specification name, and in your model you would extend the method `connection_specification_name` and return the right name.

**Comment by rzv01** [comment]
Hey @arthurnn - thanks a lot for your response. I've watched the video you linked - very informative, thanks! and it got me closer to understanding how multiple DBs should work.

Now, in your talk you mentioned that the future direction for Rails 5.1 and onward would be to have all the DB configs in the database.yml file and to initialize all connections at app startup. 

However, this means the config is now static. Would it be possible to make the app "aware" of a new DB config during runtime? 

Not sure how this would work to be honest, maybe write to the database.yml file (say, an external process which knows when a new DB is needed) and send a signal to the Rails server process? which would in turn reload the database.yml, which contains the new DB config as well.

The use case I'm thinking of is:
1. one DB per customer
2. one/few Rails app instances (each one serving all/multiple clients)
3. Rails app connects to all/multiple DBs that is assigned to it

here comes the problem:
4. customer signs up, a DB is created for him/her.

How would the Rails app know about the new DB? :)

Ideally, without having to restart the app - I guess it wouldn't be the end of the world, especially with a phased restart of a multi-threaded server, but still, not that elegant, hence me asking if a new DB config can be "injected" in the app dynamically.

Btw, sorry if I missed something and the answer to my question above is obvious!

Thanks a lot, once again! 🍻

**Comment by milep** [comment]
I'm also interested in use case where there is one DB per user/customer.

**Comment by mikecmpbll** [comment]
the database connection details don't technically need to be in database.yml, you should be able to create connection pools yourself; the code _could_ be something like this (untested):

```ruby
def connection_specification_name
  if !connection_handler.retrieve_connection_pool(Thread.current['customer_db_name'])
    spec = { adapter: "whatever", database: Thread.current['tenant'], etc: 'etc' }
    connection_handler.establish_connection(spec)
  end

  Thread.current['customer_db_name']
end
```

only looked briefly though, and i might be missing something basic there :). i assume, as we're creating a pool per database we'll end up with `pool_size *  customers` database connections.

**Comment by arthurnn** [comment]
@mikecmpbll is right. 
You can call `connection_handler.establish_connection` yourself and pass a spec. That spec needs to be a hash, which can come from anywhere, a file, db, http, etc.
So in your case, you would establish the connection, and indeed, probably better to do that during initialization, or out-of-band somehow.

**Comment by rzv01** [comment]
@mikecmpbll @arthurnn thanks a ton for your help, you guys are great :) 

I think right now i got all the info i need. I'll give it a try sometime in the next few days and see how it goes.

**Comment by milep** [comment]
Is there a thread safe way to remove idle connection pools? With some testing it looks like I just can't use ConnectionHandler's remove_connection, because it removes pool from all threads?

**Comment by njakobsen** [comment]
@arthurnn Is there a way to avoid using an instance variable to store `@connection_specification_name` in `remove_connection` (see https://github.com/rails/rails/commit/537a342a8307ebe1d5e69a7e81495f137ae571a4)? I need to dynamically swap the connection per-model, in a multi-threaded environment. Sometimes this means the same model, usually `ActiveRecord::Base`, will be connected to separate databases in separate threads. I achieve this by overriding the `connection_specification_name` method suggested in https://github.com/rails/rails/pull/24844#issuecomment-256593985 (though with a different, threadsafe implementation). However, the `remove_connection` method accesses the instance variable directly, bypassing the logic in the overridden method, and becoming un-threadsafe since the instance variable is shared across threads.

**Comment by jeremy** [comment]
Cool @njakobsen. Please do contribute a PR to no longer refer to the ivar directly. And a PR with your dynamic connection swapping would be wonderful as well! 😍

**Comment by njakobsen** [comment]
I'm definitely just crashing around, but I'll show you my monkeypatch and maybe you can let me know if there's a better way to store information about which connection the model should use on a per-thread basis. See https://github.com/culturecode/stagehand/blob/rails-5/lib/stagehand/active_record_extensions.rb#L28-L53. With some feedback, I'll see what sort of PR I can make from all of this.

**Comment by matthewd** [comment]
Do you have a use case that involves calling `remove_connection`, or are you just trying to cover the full API?

**Comment by njakobsen** [comment]
I don't have a use case for removing the connection, but if someone using my patch were to call `establish_connection` on the model, it would call `remove_connection`, so I want to make sure that it uses the overridden method to know which connection to remove based on the current thread. Otherwise I think it'll just default to AR:B's "primary" connection since no other classes are directly setting the instance variable in my monkeypatch.

**Comment by njakobsen** [comment]
I am also not happy with my solution of storing the connection spec name for each model on Thread.current. So if there are any suggestions of a better place for it, I'd happily incorporate that into a potential PR.

**Comment by mikecmpbll** [comment]
i think `establish_connection` calls [this](https://github.com/rails/rails/blob/4ca73f448187c02b5474cf90dc7c3122ee0760ff/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb#L1022) `remove_connection`

**Comment by matthewd** [comment]
If you've overwritten the getter/setter, then `@connection_specification_name` won't be defined anyway, will it?

I think it's pretty common to need to thread-scope the behaviour.. the only quick observation I'd make is that you don't necessarily need to use the setter: you could define `connection_specification_name` to directly read the top of the `ConnectionStack`, for example.

**Comment by njakobsen** [comment]
@matthewd that's an interesting idea, but I was trying to keep the changes to AR:B isolated from the inner machinations of the rest of the gem. If I could simply make it threadsafe, then any other code (including subclasses) calling `establish_connection` or `remove_connection` would continue to function as expected.

The gem I have seems to work with my changes, but at @jeremy's urging, my questions are now more about how to apply this threadsafety to a PR for Rails itself.
- Does Rails have any mechanism other than Thread.current for storing information per-thread (I see some places using `Concurrent::Map`, and `Process.pid`, so maybe something similar would be useful here)
- Do the threads ever get reaped? Or is that dependent on which webserver we're running?
- Have I missed something important?

**Comment by njakobsen** [comment]
@jeremy @matthewd I think I'm going to tackle this with a very light touch. Instead of making a bunch of decisions about how the connection should be tracked independently per-model, per-thread (which could cause issues in the general case where we don't want independent tracking per thread). I'll make a PR which simply uses the `connection_specification_name` method when removing a connection, instead of the ivar, while at the same time keeping in mind the change from this commit https://github.com/rails/rails/commit/537a342a8307ebe1d5e69a7e81495f137ae571a4. This way, an app that needs independent tracking per thread can implement it by overriding the `connection_specification_name` with whatever logic is required by their specific app.
